collect_app/src/main/java/org/odk/collect/android/http/OkHttpConnection.java

```diff
@@ -0,0 +1,352 @@
+                        multipartBuilder.addPart(MultipartBody.Part.createFormData("*isIncomplete*", "yes"));
+                        ++fileIndex; // advance over the last attachment added...
+                        break;
+                    }
+                }
+            }
+
+            MultipartBody multipartBody = multipartBuilder.build();
+
+            try {
+                OkHttpClient client = createOkHttpClient(credentials);
+                Request request = postRequest(uri, multipartBody);
+                Response response = client.newCall(request).execute();
+
+                messageParser = new ResponseMessageParser(
+                        response.toString(),
+                        response.code(),
+                        response.message());
+
+                discardEntityBytes(response);
+
+                if (response.code() != HttpURLConnection.HTTP_CREATED && response.code() != HttpURLConnection.HTTP_ACCEPTED) {
+                    return messageParser;
+                }
+            } catch (IOException e) {
+                Timber.e(e);
+                String msg = e.getMessage();
+                if (msg == null) {
+                    msg = e.toString();
+                }
+
+                throw new IOException(msg);
+            }
+        }
+
+        return messageParser;
+    }
+
+    private OkHttpClient createOkHttpClient(@Nullable HttpCredentialsInterface credentials) {
+        OkHttpClient.Builder builder = new OkHttpClient.Builder();
+
+        addCredentials(builder, credentials);
+
+        return builder
+                .connectTimeout(CONNECTION_TIMEOUT, TimeUnit.MILLISECONDS)
+                .writeTimeout(WRITE_CONNECTION_TIMEOUT, TimeUnit.MILLISECONDS)
+                .readTimeout(READ_CONNECTION_TIMEOUT, TimeUnit.MILLISECONDS)
+                .followRedirects(true)
+                .build();
+    }
+
+    private void addCredentials(OkHttpClient.Builder builder, @Nullable HttpCredentialsInterface credentials) {
+        if (credentials != null) {
+            final Map<String, CachingAuthenticator> authCache = new ConcurrentHashMap<>();
+            Credentials cred = new Credentials(credentials.getUsername(), credentials.getPassword());
+
+            BasicAuthenticator basicAuthenticator = new BasicAuthenticator(cred);
+            DigestAuthenticator digestAuthenticator = new DigestAuthenticator(cred);
+
+            DispatchingAuthenticator authenticator = new DispatchingAuthenticator.Builder()
+                    .with("digest", digestAuthenticator)
+                    .with("basic", basicAuthenticator)
+                    .build();
+
+            builder.authenticator(new CachingAuthenticatorDecorator(authenticator, authCache))
+                    .addInterceptor(new AuthenticationCacheInterceptor(authCache));
+        }
+    }
+
+    private Request getRequest(@NonNull URI uri) throws MalformedURLException {
```

Comment:
I would rename this to `buildGetRequest`. It sounds like the method gets a request. By analogy, I'd also change the ones below to `buildHeadRequest`, `buildPostRequest`.